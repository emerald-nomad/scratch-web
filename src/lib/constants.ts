export const APP_SECRET = process.env.APP_SECRET as string;
export const MONGO_URL = process.env.MONGO_URL as string;
export const API_URL = process.env.API_URL as string;