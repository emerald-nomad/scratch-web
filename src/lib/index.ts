export * from "./apolloClient"
export * from "./constants"
export * from "./test-utils"