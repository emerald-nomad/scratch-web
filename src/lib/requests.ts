import { gql } from "@apollo/client";

export const LOGIN_USER = gql`
  mutation Login($input: LoginInput!) {
    login(input: $input) {
      token
      user {
        id
        name
        username
      }
    }
  }
`;

export const SIGN_UP_USER = gql`
  mutation SignUp($input: SignUpInput!) {
    signup(input: $input) {
      token
      user {
        id
        name
        username
      }
    }
  }
`;
