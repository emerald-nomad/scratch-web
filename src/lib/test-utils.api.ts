import { ApolloServer, ServerInfo } from "apollo-server";
import getPort, { makeRange } from "get-port";
import { GraphQLClient } from "graphql-request";
import { fieldAuthorizePlugin, makeSchema } from "nexus";
import * as types from "graphql/types";
import { AuthorizationError } from "graphql/errors";

import { join } from "path";
import { mockAuthRepository, mockUserRepository } from "graphql/repositories";

const testServer = new ApolloServer({
  schema: makeSchema({
    types,
    plugins: [
      fieldAuthorizePlugin({ formatError: () => AuthorizationError() }),
    ],
    contextType: {
      module: join(process.cwd(), "src", "graphql", "./context.ts"),
      export: "Context",
    },
    outputs: {
      typegen: join(process.cwd(), "generated/nexus-typegen.ts"),
      schema: join(process.cwd(), "generated/schema.graphql"),
    },
  }),
  context: (ctx: any) => ({
    ...ctx,
    Auth: mockAuthRepository,
    User: mockUserRepository,
  }),
});

type TestContext = {
  client: GraphQLClient;
};

export function createTestContext(): TestContext {
  let ctx = {} as TestContext;
  const graphqlCtx = graphqlTestContext();

  beforeEach(async () => {
    const client = await graphqlCtx.before();
    Object.assign(ctx, {
      client,
    });
  });

  afterEach(async () => {
    await graphqlCtx.after();
  });

  return ctx;
}

function graphqlTestContext() {
  let serverInstance: ServerInfo | null = null;

  return {
    async before() {
      const port = await getPort({ port: makeRange(4000, 6000) });

      serverInstance = await testServer.listen({ port });

      console.log;

      return new GraphQLClient(`http://localhost:${port}`);
    },
    async after() {
      serverInstance?.server.close();
    },
  };
}
