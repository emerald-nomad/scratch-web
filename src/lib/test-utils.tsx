import { render, RenderOptions } from "@testing-library/react";
import { JSXElementConstructor, ReactElement } from "react";

const Providers: React.FC = ({ children }) => {
  return <> {children}</>;
};

const customRender = (
  ui: ReactElement<any, string | JSXElementConstructor<any>>,
  options: Omit<RenderOptions, "wrapper"> = {}
) => render(ui, { wrapper: Providers, ...options });

export * from "@testing-library/react";

export { customRender as render };
