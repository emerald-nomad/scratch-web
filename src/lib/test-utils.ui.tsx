import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { render, RenderOptions } from "@testing-library/react";
import { JSXElementConstructor, ReactElement } from "react";

export type MockResponse = MockedResponse<Record<string, any>>;
interface Params {
  options?: Omit<RenderOptions, "wrapper">;
  mocks?: MockResponse[];
}

const customRender = (
  ui: ReactElement<any, string | JSXElementConstructor<any>>,
  params?: Params
) => {
  const Wrapper: React.FC = ({ children }) => {
    return (
      <MockedProvider mocks={params?.mocks ?? []}>{children}</MockedProvider>
    );
  };

  return render(ui, { wrapper: Wrapper, ...(params?.options ?? {}) });
};

export * from "@testing-library/react";

export { customRender as render };
