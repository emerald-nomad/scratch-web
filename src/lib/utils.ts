import { UseFormSetError } from "react-hook-form";

export const setGraphqlErrors = ({
  errors,
  setError,
}: {
  errors: { [key: string]: string };
  setError: UseFormSetError<any>;
}) => {
  for (const prop in errors) {
    setError(prop, { message: errors[prop] });
  }
};
