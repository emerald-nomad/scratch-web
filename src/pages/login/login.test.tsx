/**
 * @jest-environment jsdom
 */

import React from "react";
import userEvent from "@testing-library/user-event";
import Login, { LOGIN_USER } from "./index";
import { cleanup, MockResponse, render, screen, act } from "lib";

const testPassword = "12345";
const failingUsername = "Error";

const mocks: MockResponse[] = [
  {
    request: {
      query: LOGIN_USER,
      variables: {
        input: {
          username: failingUsername,
          password: testPassword,
        },
      },
    },
    error: {
      name: "Test",
      message: "Test",
    },
  },
  {
    request: {
      query: LOGIN_USER,
      variables: {
        input: {
          username: "testuser",
          password: testPassword,
        },
      },
    },
    result: {
      data: {
        id: "1",
        name: "Test User",
        username: "testuser",
      },
    },
  },
];

describe("Login Page", () => {
  beforeEach(() => {
    render(<Login />, { mocks });
  });

  afterEach(() => {
    cleanup();
  });

  it("should have login form", async () => {
    const loginForm = await screen.getByLabelText("Login Form");

    expect(loginForm).toMatchSnapshot();
  });

  it("should have banner section", async () => {
    const banner = await screen.getByLabelText("Banner");

    expect(banner).toMatchSnapshot();
  });

  it("should have a link to signup page", async () => {
    const signupLink = await screen.getByLabelText("Sign Up Link");

    expect(signupLink).toMatchSnapshot();
  });

  it("should show errors for submitting form with empty inputs", async () => {
    const submitBtn = screen.getByText("Login");

    act(() => {
      userEvent.click(submitBtn);
    });

    expect(await screen.findByText("Username is required"));
    expect(await screen.findByText("Password is required"));
    expect(submitBtn).toMatchSnapshot();
  });

  it.todo("should display error for bad username/password combination");

  it.todo("should redirect to 'feed' page on successful login");
});
