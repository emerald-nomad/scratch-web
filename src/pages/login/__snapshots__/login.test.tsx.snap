// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should have a link to signup page 1`] = `
<div
  aria-label="Sign Up Link"
>
  <p>
    New to Scratch?
  </p>
  <button
    class="btn btn--text"
    type="button"
  >
    Sign Up!
  </button>
</div>
`;

exports[`Login Page should have banner section 1`] = `
<div
  aria-label="Banner"
  class="banner"
>
  <div
    style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
  >
    <noscript />
    <img
      alt="Logo"
      decoding="async"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
    />
  </div>
  <h1
    aria-label="Banner Header"
    class="banner__header"
  >
    Welcome Back!
  </h1>
  <div
    style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
  >
    <noscript />
    <img
      alt="Banner Image"
      decoding="async"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
    />
  </div>
</div>
`;

exports[`Login Page should have login form 1`] = `
<form
  aria-label="Login Form"
>
  <div
    class="container"
  >
    <div
      class="container__content"
    >
      <label
        for="username"
      >
        Username
      </label>
      Forgot Password
    </div>
    <input
      name="username"
      type="text"
      value=""
    />
    <p
      aria-label="Username error message"
    />
  </div>
  <div
    class="container"
  >
    <div
      class="container__content"
    >
      <label
        for="password"
      >
        Password
      </label>
    </div>
    <input
      name="password"
      type="text"
      value=""
    />
    <p
      aria-label="Password error message"
    />
  </div>
  <button
    class="btn btn--solid"
    type="submit"
  >
    Login
  </button>
</form>
`;

exports[`Login Page should show errors for submitting form with empty inputs 1`] = `
<button
  class="btn btn--solid"
  type="submit"
>
  Login
</button>
`;
